quinn-proto/src/transport_parameters.rs

```diff
@@ -295,68 +363,101 @@ impl From<UnexpectedEnd> for Error {
 impl TransportParameters {
     /// Encode `TransportParameters` into buffer
     pub fn write<W: BufMut>(&self, w: &mut W) {
-        macro_rules! write_params {
-            {$($(#[$doc:meta])* $name:ident ($code:expr) = $default:expr,)*} => {
-                $(
-                    if self.$name.0 != $default {
-                        w.write_var($code);
-                        w.write(VarInt::try_from(self.$name.size()).unwrap());
-                        w.write(self.$name);
+        for id in self
+            .write_order
+            .as_ref()
+            .unwrap_or(&DEFAULT_TRANSPORT_PARAMETERS_ORDER)
+        {
+            match id {
+                TransportParameterId::ReservedTransportParameter => {
+                    if let Some(param) = self.grease_transport_parameter {
+                        param.write(w);
                     }
-                )*
-            }
-        }
-        apply_params!(write_params);
-
-        if let Some(param) = self.grease_transport_parameter {
-            param.write(w);
-        }
-
-        if let Some(ref x) = self.stateless_reset_token {
-            w.write_var(0x02);
-            w.write_var(16);
-            w.put_slice(x);
-        }
-
-        if self.disable_active_migration {
-            w.write_var(0x0c);
-            w.write_var(0);
-        }
-
-        if let Some(x) = self.max_datagram_frame_size {
-            w.write_var(0x20);
-            w.write_var(x.size() as u64);
-            w.write(x);
-        }
-
-        if let Some(ref x) = self.preferred_address {
-            w.write_var(0x000d);
-            w.write_var(x.wire_size() as u64);
-            x.write(w);
-        }
-
-        for &(tag, cid) in &[
-            (0x00, &self.original_dst_cid),
-            (0x0f, &self.initial_src_cid),
-            (0x10, &self.retry_src_cid),
-        ] {
-            if let Some(ref cid) = *cid {
-                w.write_var(tag);
-                w.write_var(cid.len() as u64);
-                w.put_slice(cid);
+                }
+                TransportParameterId::StatelessResetToken => {
+                    if let Some(ref x) = self.stateless_reset_token {
+                        w.write_var(TransportParameterId::StatelessResetToken as u64);
```

Comment:
style: prefer `id as u64` here and below so there's less risk of copy-paste errors in the future.